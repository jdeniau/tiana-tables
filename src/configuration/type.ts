import { ConnectionObject } from '../component/Connection/types';

export type Configuration = {
  version: 1;
  theme: string;
  connections: Record<string, EncryptedConnectionObject>;
};

export type ConnectionAppState = {
  isActive: boolean; // TODO not handled yet
  activeDatabase: string;
  openedTable: string; // TODO not handled yet
};

export type EncryptedConnectionObject = {
  password: string;
  appState?: ConnectionAppState;
} & Omit<ConnectionObject, 'password'>;

export type EncryptedConfiguration = {
  connections: Record<string, EncryptedConnectionObject>;
} & Omit<Configuration, 'connections'>;
