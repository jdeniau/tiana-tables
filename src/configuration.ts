import { dialog, safeStorage } from 'electron';
import path from 'node:path';
import fs from 'node:fs';
import envPaths from 'env-paths';
import { ConnectionObject } from './component/Connection';

export type Configuration = {
  version: 1;
  connections: Record<string, EncryptedConnectionObject>;
};

type EncryptedConnectionObject = {
  password: string;
} & Omit<ConnectionObject, 'password'>;

type EncryptedConfiguration = {
  connections: Record<string, EncryptedConnectionObject>;
} & Omit<Configuration, 'connections'>;

const envPath = envPaths('FuzzyPotato', { suffix: '' });
const dataFilePath = path.resolve(envPath.config, 'config.json');

function encryptConnection(
  connection: ConnectionObject
): EncryptedConnectionObject {
  return {
    ...connection,
    password: safeStorage.encryptString(connection.password).toString('base64'),
  };
}

function decryptConnection(
  connection: EncryptedConnectionObject
): ConnectionObject {
  return {
    ...connection,
    password: safeStorage.decryptString(
      Buffer.from(connection.password, 'base64')
    ),
  };
}

export function readConfigurationFile(): null | Configuration {
  if (!fs.existsSync(dataFilePath)) {
    return null;
  }

  const dataString = fs.readFileSync(dataFilePath, 'utf-8');

  let config = JSON.parse(dataString) as EncryptedConfiguration;

  return {
    ...config,
    connections: Object.fromEntries(
      Object.entries(config.connections).map(([name, connection]) => [
        name,
        decryptConnection(connection),
      ])
    ),
  };
}
export function addConnectionToConfig(
  event: Electron.IpcMainInvokeEvent,
  name: string,
  connection: ConnectionObject
): void {
  let config = readConfigurationFile();

  if (!config) {
    config = {
      version: 1,
      connections: {},
    };
  }

  if (!config.connections) {
    config.connections = {};
  }

  config.connections[name] = encryptConnection(connection);

  fs.mkdirSync(envPath.config, { recursive: true });
  fs.writeFile(
    dataFilePath,
    JSON.stringify(config, null, 2),
    'utf-8',
    (err) => {
      if (err) {
        dialog.showErrorBox('Error', err.message);
      }
    }
  );
}
