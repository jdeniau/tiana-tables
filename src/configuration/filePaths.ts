import { app } from 'electron';
import { existsSync, mkdirSync } from 'node:fs';
import path, { dirname, resolve } from 'node:path';
import { isDevApp } from '../main-process/helpers';

export function getConfigurationPath() {
  const configDir = resolve(app.getPath('userData'), 'config');
  const configPath = resolve(configDir, 'config.json');

  return configPath;
}

export function getConfigurationFolder() {
  return dirname(getConfigurationPath());
}

export function createConfigurationFolderIfNotExists() {
  const configurationPah = getConfigurationPath();

  if (!existsSync(dirname(configurationPah))) {
    mkdirSync(dirname(configurationPah), { recursive: true });
  }
}

export function getLogPath() {
  return path.join(
    app.getPath('userData'),
    'logs',
    isDevApp() ? 'dev.log' : 'main.log'
  );
}

export function getLogFolder() {
  return dirname(getLogPath());
}
