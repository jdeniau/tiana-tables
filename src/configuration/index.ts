import { dialog, safeStorage } from 'electron';
import { resolve } from 'node:path';
import { existsSync, mkdirSync, readFileSync, writeFile } from 'node:fs';
import envPaths from 'env-paths';
import { ConnectionObject } from '../component/Connection/types';
import { DEFAULT_THEME } from '../theme';
import {
  Configuration,
  EncryptedConnectionObject,
  EncryptedConfiguration,
  ConnectionAppState,
} from './type';

// TODO use app.getPath('userData') to store the configuration file instead of env-paths
const envPath = envPaths('TianaTables', { suffix: '' });
const dataFilePath = resolve(envPath.config, 'config.json');

console.log('Configuration file path:', dataFilePath);

function getBaseConfig(): Configuration {
  return {
    version: 1,
    theme: DEFAULT_THEME.name,
    connections: {},
  };
}

function encryptConnection(
  connection: ConnectionObject
): EncryptedConnectionObject {
  return {
    ...connection,
    password: safeStorage.encryptString(connection.password).toString('base64'),
  };
}

function decryptConnection(
  connection: EncryptedConnectionObject
): ConnectionObject {
  return {
    ...connection,
    password: safeStorage.decryptString(
      Buffer.from(connection.password, 'base64')
    ),
  };
}

export function getConfiguration(): Configuration {
  if (!existsSync(dataFilePath)) {
    return getBaseConfig();
  }

  const dataString = readFileSync(dataFilePath, 'utf-8');

  if (!dataString) {
    return getBaseConfig();
  }

  const config = JSON.parse(dataString) as EncryptedConfiguration;

  return {
    ...config,
    connections: Object.fromEntries(
      Object.entries(config.connections ?? {}).map(([name, connection]) => [
        name,
        decryptConnection(connection),
      ])
    ),
  };
}

function writeConfiguration(config: Configuration): void {
  const encryptedConfig = {
    ...config,
    connections: Object.fromEntries(
      Object.entries(config.connections).map(([name, connection]) => [
        name,
        encryptConnection(connection),
      ])
    ),
  };

  mkdirSync(envPath.config, { recursive: true });
  writeFile(
    dataFilePath,
    JSON.stringify(encryptedConfig, null, 2),
    'utf-8',
    (err) => {
      if (err) {
        dialog.showErrorBox('Error', err.message);
      }
    }
  );
}

export function addConnectionToConfig(connection: ConnectionObject): void {
  const config = getConfiguration() ?? getBaseConfig();

  if (!config.connections) {
    config.connections = {};
  }

  config.connections[connection.name] = connection;

  writeConfiguration(config);
}

export function changeTheme(theme: string): void {
  const config = getConfiguration() ?? getBaseConfig();
  config.theme = theme;

  writeConfiguration(config);
}

export function updateConnectionState<K extends keyof ConnectionAppState>(
  connectionName: string,
  key: K,
  value: ConnectionAppState[K]
) {
  const config = getConfiguration();

  const connection = config.connections[connectionName];

  if (!connection) {
    return;
  }

  if (!connection.appState) {
    connection.appState = {
      isActive: false,
      activeDatabase: '',
      openedTable: '',
    };
  }

  connection.appState[key] = value;

  writeConfiguration(config);
}

export const testables = { getBaseConfig };
