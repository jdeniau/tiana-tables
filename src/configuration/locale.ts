export const DEFAULT_LOCALE = 'en';

export const LOCALE_LIST = ['en', 'fr'];
