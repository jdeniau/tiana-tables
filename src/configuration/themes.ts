import * as dracula from '../renderer/theme/dracula.json';
import * as visualStudio from '../renderer/theme/visualStudio.json';
import * as active4d from '../renderer/theme/active4d.json';

export interface TmTheme {
  readonly name: string;
  readonly author?: string;
  readonly settings: TmThemeSetting[];
}

const THEME_LIST_AS_ARRAY: TmTheme[] = [
  dracula,
  visualStudio,
  active4d,
] as const;

const DARK_THEME_LIST_NAME = [dracula.name];

type Theme = (typeof THEME_LIST_AS_ARRAY)[number];

export const THEME_LIST: Record<string, Theme> = {};
THEME_LIST_AS_ARRAY.forEach((t) => {
  THEME_LIST[t.name] = t;
});

export const DEFAULT_THEME = THEME_LIST_AS_ARRAY[0];

export function isDarkTheme(theme: Theme): boolean {
  return DARK_THEME_LIST_NAME.includes(theme.name);
}

export interface TmThemeSetting {
  readonly scope?: string | string[];
  readonly settings: Record<string, string | undefined>;
}

export interface TmThemeScopedSetting extends TmThemeSetting {
  readonly scope: string | string[];
}

export interface TmThemeGlobalSetting extends TmThemeSetting {
  readonly scope: undefined;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isScopedSetting(o: { scope?: any; settings?: any }): boolean {
  return o.scope && o.settings;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
export function isUnscopedSetting(o: { scope?: any; settings?: any }): boolean {
  return !o.scope && o.settings;
}
