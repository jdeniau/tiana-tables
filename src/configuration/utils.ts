export function slugify(name: string): string {
  return name
    .toLowerCase()
    .replace(/ /g, '-')
    .replace(/[^a-z0-9-]/g, '')
    .replace(/-{1,}/g, '-')
    .replace(/-$/, '')
    .replace(/^-/, '');
}
