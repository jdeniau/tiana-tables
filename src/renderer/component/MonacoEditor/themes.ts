import { editor } from 'monaco-editor';
import {
  TmTheme,
  isDarkTheme,
  isUnscopedSetting,
} from '../../../configuration/themes';

// Taken from https://github.com/brijeshb42/monaco-themes/blob/2bf2a1f3a3a1e43e0b91c4a67376dd4051b5055d/src/index.js
// but monaco-themes needs a pList file to work
// It alse guess if the theme is dark or light, but we already have that information

function parseColor(color: string): string | null {
  if (!color.length) {
    return null;
  }

  if (color.length == 4) {
    color = color.replace(/[a-fA-F\d]/g, '$&$&');
  }

  if (color.length == 7) {
    return color;
  }

  if (color.length == 9) {
    return color; // substr(0, 7);
  } else {
    if (!color.match(/^#(..)(..)(..)(..)$/)) {
      throw new Error(`can't parse color ${color}`);
    }

    const rgba: Array<number> | undefined = color
      .match(/^#(..)(..)(..)(..)$/)
      ?.slice(1)
      .map(function (c) {
        return parseInt(c, 16);
      });

    if (!rgba) {
      throw new Error(`can't parse color ${color}`);
    }

    // @ts-expect-error string / number assignation issue, but will be joined the line after
    rgba[3] = (rgba[3] / 0xff).toPrecision(2);

    return 'rgba(' + rgba.join(', ') + ')';
  }
}

/* Mapped from vscode/src/vs/workbench/services/themes/electron-browser/themeCompatibility.ts */
const COLOR_MAP = [
  {
    tm: 'foreground',
    mn: 'editor.foreground',
  },
  {
    tm: 'background',
    mn: 'editor.background',
  },
  // {
  //   tm: 'foreground',
  //   mn: 'editorSuggestWidget.foreground',
  // },
  // {
  //   tm: 'background',
  //   mn: 'editorSuggestWidget.background',
  // },
  {
    tm: 'selection',
    mn: 'editor.selectionBackground',
  },
  {
    tm: 'inactiveSelection',
    mn: 'editor.inactiveSelectionBackground',
  },
  {
    tm: 'selectionHighlightColor',
    mn: 'editor.selectionHighlightBackground',
  },
  {
    tm: 'findMatchHighlight',
    mn: 'editor.findMatchHighlightBackground',
  },
  {
    tm: 'currentFindMatchHighlight',
    mn: 'editor.findMatchBackground',
  },
  {
    tm: 'hoverHighlight',
    mn: 'editor.hoverHighlightBackground',
  },
  {
    tm: 'wordHighlight',
    mn: 'editor.wordHighlightBackground',
  },
  {
    tm: 'wordHighlightStrong',
    mn: 'editor.wordHighlightStrongBackground',
  },
  {
    tm: 'findRangeHighlight',
    mn: 'editor.findRangeHighlightBackground',
  },
  {
    tm: 'findMatchHighlight',
    mn: 'peekViewResult.matchHighlightBackground',
  },
  {
    tm: 'referenceHighlight',
    mn: 'peekViewEditor.matchHighlightBackground',
  },
  {
    tm: 'lineHighlight',
    mn: 'editor.lineHighlightBackground',
  },
  {
    tm: 'rangeHighlight',
    mn: 'editor.rangeHighlightBackground',
  },
  {
    tm: 'caret',
    mn: 'editorCursor.foreground',
  },
  {
    tm: 'invisibles',
    mn: 'editorWhitespace.foreground',
  },
  {
    tm: 'guide',
    mn: 'editorIndentGuide.background',
  },
  {
    tm: 'activeGuide',
    mn: 'editorIndentGuide.activeBackground',
  },
  {
    tm: 'selectionBorder',
    mn: 'editor.selectionHighlightBorder',
  },
];

const ansiColorMap = [
  'ansiBlack',
  'ansiRed',
  'ansiGreen',
  'ansiYellow',
  'ansiBlue',
  'ansiMagenta',
  'ansiCyan',
  'ansiWhite',
  'ansiBrightBlack',
  'ansiBrightRed',
  'ansiBrightGreen',
  'ansiBrightYellow',
  'ansiBrightBlue',
  'ansiBrightMagenta',
  'ansiBrightCyan',
  'ansiBrightWhite',
];

ansiColorMap.forEach((color) => {
  COLOR_MAP.push({
    tm: color,
    mn: 'terminal.' + color,
  });
});
export function convertTextmateThemeToMonaco(
  rawData: TmTheme
): editor.IStandaloneThemeData {
  const firstSettings = rawData.settings[0];

  if (!isUnscopedSetting(firstSettings)) {
    throw new Error('Expected global settings in first position');
  }
  const globalSettings = firstSettings.settings;

  const rules: editor.ITokenThemeRule[] = [];

  rawData.settings.forEach((setting) => {
    if (!setting.settings) {
      return;
    }

    let scopes;

    if (typeof setting.scope === 'string') {
      scopes = setting.scope
        .replace(/^[,]+/, '')
        .replace(/[,]+$/, '')
        .split(',');
    } else if (Array.isArray(setting.scope)) {
      scopes = setting.scope;
    } else {
      scopes = [''];
    }

    const rule: Partial<editor.ITokenThemeRule> = {};
    const settings = setting.settings;

    if (settings.foreground) {
      const parsed = parseColor(settings.foreground);
      if (parsed) {
        rule.foreground = parsed.toLowerCase().replace('#', '');
      }
    }

    if (settings.background) {
      const parsed = parseColor(settings.background);
      if (parsed) {
        rule.background = parsed.toLowerCase().replace('#', '');
      }
    }

    if (settings.fontStyle && typeof settings.fontStyle === 'string') {
      rule.fontStyle = settings.fontStyle;
    }

    scopes.forEach((scope) => {
      if (!scope || !Object.keys(rule).length) {
        return;
      }
      const r = Object.assign({}, rule, {
        token: scope.trim(),
      });
      rules.push(r);
    });
  });

  const globalColors: editor.IColors = {};

  /* More properties to be added */
  COLOR_MAP.forEach((obj) => {
    const tmGlobalSetting = globalSettings[obj.tm];

    if (tmGlobalSetting) {
      const parsed = parseColor(tmGlobalSetting);

      if (parsed) {
        globalColors[obj.mn] = parsed;
      }
    }
  });

  const base = isDarkTheme(rawData) ? 'vs-dark' : 'vs';

  return {
    base,
    inherit: true,
    rules: rules,
    colors: globalColors,
  };
}
