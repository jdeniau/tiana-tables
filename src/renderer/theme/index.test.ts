import { describe, expect, test } from 'vitest';
import { THEME_LIST } from '../../configuration/themes';
import {
  background,
  constantLanguageForeground,
  constantLanguageNullForeground,
  constantNumericForeground,
  foreground,
  selection,
  stringForeground,
  supportTypeForeground,
} from '.';

describe.each(Object.values(THEME_LIST))('style helpers', (theme) => {
  test('foreground', () => {
    expect(foreground({ theme })).toMatchSnapshot(theme.name);
  });

  test('background', () => {
    expect(background({ theme })).toMatchSnapshot(theme.name);
  });

  test('selection', () => {
    expect(selection({ theme })).toMatchSnapshot(theme.name);
  });

  test('stringForeground', () => {
    expect(stringForeground({ theme })).toMatchSnapshot(theme.name);
  });

  test('supportTypeForeground', () => {
    expect(supportTypeForeground({ theme })).toMatchSnapshot(theme.name);
  });

  test('constantLanguageForeground', () => {
    expect(constantLanguageForeground({ theme })).toMatchSnapshot(theme.name);
  });

  test('constantNumericForeground', () => {
    expect(constantNumericForeground({ theme })).toMatchSnapshot(theme.name);
  });

  test('constantLanguageNullForeground', () => {
    expect(constantLanguageNullForeground({ theme })).toMatchSnapshot(
      theme.name
    );
  });
});
