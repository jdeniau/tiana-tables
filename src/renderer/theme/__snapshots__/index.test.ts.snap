// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`style helpers > background > Active4D 1`] = `"#FFFFFF"`;

exports[`style helpers > background > Base16 Eva 1`] = `"#2a3b4d"`;

exports[`style helpers > background > Base16 Unikitty Light 1`] = `"#ffffff"`;

exports[`style helpers > background > Dracula 1`] = `"#282a36"`;

exports[`style helpers > background > Night Owl 1`] = `"#002B36"`;

exports[`style helpers > background > Visual Studio 1`] = `"#FFFFFF"`;

exports[`style helpers > constantLanguageForeground > Active4D 1`] = `"#A535AE"`;

exports[`style helpers > constantLanguageForeground > Base16 Eva 1`] = `"#ff9966"`;

exports[`style helpers > constantLanguageForeground > Base16 Unikitty Light 1`] = `"#d65407"`;

exports[`style helpers > constantLanguageForeground > Dracula 1`] = `"#bd93f9"`;

exports[`style helpers > constantLanguageForeground > Night Owl 1`] = `"#82AAFF"`;

exports[`style helpers > constantLanguageForeground > Visual Studio 1`] = `"#000000"`;

exports[`style helpers > constantLanguageNullForeground > Active4D 1`] = `"#A535AE"`;

exports[`style helpers > constantLanguageNullForeground > Base16 Eva 1`] = `"#ff9966"`;

exports[`style helpers > constantLanguageNullForeground > Base16 Unikitty Light 1`] = `"#d65407"`;

exports[`style helpers > constantLanguageNullForeground > Dracula 1`] = `"#bd93f9"`;

exports[`style helpers > constantLanguageNullForeground > Night Owl 1`] = `"#82AAFF"`;

exports[`style helpers > constantLanguageNullForeground > Visual Studio 1`] = `"#000000"`;

exports[`style helpers > constantNumericForeground > Active4D 1`] = `"#A8017E"`;

exports[`style helpers > constantNumericForeground > Base16 Eva 1`] = `"#ff9966"`;

exports[`style helpers > constantNumericForeground > Base16 Unikitty Light 1`] = `"#d65407"`;

exports[`style helpers > constantNumericForeground > Dracula 1`] = `"#bd93f9"`;

exports[`style helpers > constantNumericForeground > Night Owl 1`] = `"#F78C6C"`;

exports[`style helpers > constantNumericForeground > Visual Studio 1`] = `"#A31515"`;

exports[`style helpers > foreground > Active4D 1`] = `"#3B3B3B"`;

exports[`style helpers > foreground > Base16 Eva 1`] = `"#9fa2a6"`;

exports[`style helpers > foreground > Base16 Unikitty Light 1`] = `"#6c696e"`;

exports[`style helpers > foreground > Dracula 1`] = `"#f8f8f2"`;

exports[`style helpers > foreground > Night Owl 1`] = `"#839496"`;

exports[`style helpers > foreground > Visual Studio 1`] = `"#000000"`;

exports[`style helpers > selection > Active4D 1`] = `"#BAD6FD"`;

exports[`style helpers > selection > Base16 Eva 1`] = `"#4b6988"`;

exports[`style helpers > selection > Base16 Unikitty Light 1`] = `"#c4c3c5"`;

exports[`style helpers > selection > Dracula 1`] = `"#44475a"`;

exports[`style helpers > selection > Night Owl 1`] = `"#073642"`;

exports[`style helpers > selection > Visual Studio 1`] = `"#9DA7C3"`;

exports[`style helpers > stringForeground > Active4D 1`] = `"#666666"`;

exports[`style helpers > stringForeground > Base16 Eva 1`] = `"#66ff66"`;

exports[`style helpers > stringForeground > Base16 Unikitty Light 1`] = `"#17ad98"`;

exports[`style helpers > stringForeground > Dracula 1`] = `"#f1fa8c"`;

exports[`style helpers > stringForeground > Night Owl 1`] = `"#addb67"`;

exports[`style helpers > stringForeground > Visual Studio 1`] = `"#A31515"`;

exports[`style helpers > supportTypeForeground > Active4D 1`] = `"#A535AE"`;

exports[`style helpers > supportTypeForeground > Base16 Eva 1`] = `"#4b8f77"`;

exports[`style helpers > supportTypeForeground > Base16 Unikitty Light 1`] = `"#149bda"`;

exports[`style helpers > supportTypeForeground > Dracula 1`] = `"#66d9ef"`;

exports[`style helpers > supportTypeForeground > Night Owl 1`] = `"#addb67"`;

exports[`style helpers > supportTypeForeground > Visual Studio 1`] = `"#000000"`;
